tests/e2e_shared.rs

```diff
@@ -0,0 +1,477 @@
+#![cfg(feature = "_e2e_tests")]
+use std::time::Duration;
+use std::{collections::HashMap, path::PathBuf};
+
+use anyhow::{ensure, Context as _};
+use futures::StreamExt;
+use helpers::HTTP_CLIENT_IMAGE_REF;
+use wadm_types::api::StatusType;
+
+mod e2e;
+mod helpers;
+
+use e2e::{assert_status, check_components, check_providers, ClientInfo, ExpectedCount};
+
+use crate::{
+    e2e::check_status,
+    helpers::{HELLO_IMAGE_REF, HTTP_SERVER_IMAGE_REF},
+};
+
+const MANIFESTS_PATH: &str = "tests/fixtures/manifests/shared";
+const DOCKER_COMPOSE_FILE: &str = "tests/docker-compose-e2e_shared.yaml";
+
+const SHARED_COMPONENTS_LATTICE: &str = "shared_components";
+const SHARED_PROVIDERS_LATTICE: &str = "shared_providers";
+const INVALID_TEST_LATTICE: &str = "shared_invalid";
+
+#[cfg(feature = "_e2e_tests")]
+#[tokio::test(flavor = "multi_thread")]
+async fn run_shared_component_tests() {
+    use futures::FutureExt;
+
+    let root_dir =
+        PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").expect("Unable to find repo root"));
+    let manifest_dir = root_dir.join(MANIFESTS_PATH);
+    let compose_file = root_dir.join(DOCKER_COMPOSE_FILE);
+
+    let mut client_info = ClientInfo::new(manifest_dir, compose_file).await;
+    client_info
+        .add_ctl_client(SHARED_COMPONENTS_LATTICE, None)
+        .await;
+    client_info.add_wadm_client(SHARED_COMPONENTS_LATTICE).await;
+    client_info
+        .add_ctl_client(SHARED_PROVIDERS_LATTICE, None)
+        .await;
+    client_info.add_wadm_client(SHARED_PROVIDERS_LATTICE).await;
+    client_info.add_ctl_client(INVALID_TEST_LATTICE, None).await;
+    client_info.add_wadm_client(INVALID_TEST_LATTICE).await;
+    client_info.launch_wadm().await;
+
+    // Wait for the first event on the lattice prefix before we start deploying and checking
+    // statuses. Wadm can absolutely handle hosts starting before you start the wadm process, but the first event
+    // on the lattice will initialize the lattice monitor and for the following test we quickly assert things.
+    let mut sub = client_info
```

Comment:
Nit: Should we pull out the setup stuff here (which I think you copied) into a common helper for the tests?

Reply:
Agreed! I added an issue for this https://github.com/wasmCloud/wadm/issues/456